### PR TITLE
ptrace emulation improvements, round 2

### PR DESCRIPTION
**Pull request by Keno**
As usual, see commit messages for details. Also, don't worry, @rocallahan, I'm almost through my backlog of local changes (3-4 commits left), so the PR storm will be over soon.

### CODE REVIEW

**Comment by rocallahan** [comment]
There's a test failure here in `ptrace_tls`

File: src/Task.cc [hunk 1]
```diff
@@ -1074,6 +1094,29 @@ void Task::set_thread_area(remote_ptr<struct user_desc> tls) {
...
+  // We rely on the fact that user_desc is word-size-independent.
+  auto desc = read_mem(tls);
+  rr::set_thread_area(thread_areas_, desc);
+}
+
+int Task::emulate_set_thread_area(int idx, struct ::user_desc desc) {
+  errno = 0;
+  fallible_ptrace(PTRACE_SET_THREAD_AREA, idx, &desc);
+  if (errno != 0)
+    return errno;
```
Comment (rocallahan) [inline src/Task.cc]: {}

File: src/record_syscall.cc [hunk 2]
```diff
@@ -2776,8 +2852,9 @@ static Switchable rec_prepare_syscall_arch(RecordTask* t,
           1, ParamSize::from_syscall_result<int>((size_t)t->regs().arg2()));
       return (GRND_NONBLOCK & t->regs().arg3()) ? PREVENT_SWITCH : ALLOW_SWITCH;
 
+    case Arch::get_thread_area:
     case Arch::set_thread_area:
-      syscall_state.reg_parameter<typename Arch::user_desc>(1, IN_OUT);
```
Comment (rocallahan) [inline src/record_syscall.cc]: Why did you change this?

File: src/RecordTask.cc [hunk 3]
```diff
@@ -1588,9 +1588,15 @@ void RecordTask::set_tid_addr(remote_ptr<int> tid_addr) {
 }
 
 void RecordTask::tgkill(int sig) {
+  LOG(debug) << "Sending " << sig << " to tid " << tid;
   ASSERT(this, 0 == syscall(SYS_tgkill, real_tgid(), tid, sig));
 }
 
+void RecordTask::kill_if_alive() {
+  if (!is_dying())
+    tgkill(SIGKILL);
```
Comment (rocallahan) [inline src/RecordTask.cc]: {}

**Comment by rocallahan** [comment]
So, basically looks good apart from the test failure.

File: src/record_syscall.cc (hunk 2 above)
Comment (Keno) [inline src/record_syscall.cc]: `::user_desc` is used in most other places (such that the same method can be used for both archs), so it just seemed more uniform. Happy to change it back though.

**Comment by Keno** [comment]
Not seeing the test failure locally as usual. Will take a look at the buildbot logs.

**Comment by Keno** [comment]
Ah, not an rr bug, the test fails outside of rr as well. As best as I can tell, the `PTRACE_ARCH_PRCTL` is not taking effect for some reason. Probably an already fixed kernel bug,
I suspect https://github.com/torvalds/linux/commit/d47b50e7a111bb7a56fb1c974728b56209d7f515 may be  the fix. Will build a 4.6 kernel and see if I'm right.

**Comment by Keno** [comment]
I locally built both 4.6 and mainline 4.4 (the buildbot is at ubuntu's fork of 4.4), but was unable to reproduce this problem.

**Comment by rocallahan** [comment]
Do you want me to look into it by logging into the ci box?

**Comment by Keno** [comment]
@khuey gave me access to the ci box, so I'm looking into it. The failure seems pretty straightforward: The PTRACE_ARCH_PRCTL is not taking effect. I've just not beed able to reproduce that behavior outside the ci box at all.

**Comment by rocallahan** [comment]
I think it's fine to modify the test to tolerate that without digging down all the way to find the root cause.

**Comment by khuey** [comment]
Some weird KVM interaction, perhaps?

**Comment by Keno** [comment]
Possible. Do you have KVM set up locally? Can you run the ptrace_tls test form this PR and see if it runs?

**Comment by khuey** [comment]
I can try that later today.

**Comment by rocallahan** [comment]
I'll try it now.

**Comment by rocallahan** [comment]
Yes, `ptrace_tls` fails in my KVM running guest kernel 4.4.0.

**Comment by Keno** [comment]
> Yes, ptrace_tls fails in my KVM running guest kernel 4.4.0.

Can't reproduce :(. Can you post host kernel version and guest kernel .config?

**Comment by Keno** [comment]
Actually hold on, this may be a libc difference not a kernel difference. I just uploaded my binary to the buildbot and it runs fine.

**Comment by Keno** [comment]
Spoke too soon. I had built it as a static binary for testing (KVM only). But if I build it as a static binary on the buildbot it works fine as well (as does the dynamic binary I downloaded from the buildbot and ran on my development machine).

**Comment by Keno** [comment]
Ok, I think I'm on to it. May have misdiagnosed the issue (not the effect, but the root cause).

**Comment by Keno** [comment]
Well, good news, bad news and good news. Good news is I found the libc problem that was causing the issue. Bad news it still fails on the build bot. Good news, I can confirm the new failure mode, so the kernel bisection tool I wrote to track down the earlier failure should figure this one out tout de suite.

**Comment by Keno** [comment]
Ok, bisect shows the commit that fixed this is https://github.com/torvalds/linux/commit/731e33e39a5b95ad77017811b3ced32ecf9dc666 . I'll address the stylistic comments, and adjust the test to allow failure on kernels < 4.7.

BTW, here's the script I used, for my own future reference, and in case others may have to do a kernel bisect. Was super useful: https://gist.github.com/Keno/6e506f51a6ab4b606dbfccc1159d94c9

**Comment by Keno** [comment]
For posterity, here's what the bug is:
```
ptrace.c:
static int putreg(struct task_struct *child,
		  unsigned long offset, unsigned long value)
[snip]
		if (child->thread.fsbase != value)
			return do_arch_prctl(child, ARCH_SET_FS, value);
		return 0;
```
The problem is that `fsbase` (called `fs` back in relevant kernel versions) wasn't actually the right thing to check for, because it could have used the segment table optimization. However, if it does that it sets `fsbase = 0`, so when you try to reset the actual fs_base to 0, it thinks it's already there and the change doesn't go through. Now that I think about it, this is a bit of a special case, so maybe I'll just hack that into rr.

**Comment by Keno** [comment]
@rocallahan should be ready

File: src/Task.cc [hunk 4]
```diff
@@ -487,6 +487,33 @@ void Task::on_syscall_exit_arch(int syscallno, const Registers& regs) {
...
+          int code = (int)regs.arg4();
+          switch (code) {
+            case ARCH_GET_FS:
+            case ARCH_GET_GS:
+              break;
+            case ARCH_SET_FS:
+            case ARCH_SET_GS: {
+              Registers r = tracee->regs();
+              if (regs.arg3() == 0) {
+                // Work around a kernel bug in pre-3.7 kernels, where setting
```
Comment (rocallahan) [inline src/Task.cc]: This should be "4.7"

Comment (Keno) [inline src/Task.cc]: Hah, yeah. Have made that mistake more than once. Fixed locally, will push together with anything else once you're done reviewing.

Comment (rocallahan) [inline src/Task.cc]: That's all I noticed. Just repush and I'll merge it.

Comment (Keno) [inline src/Task.cc]: Done

**Comment by rocallahan** [comment]
Thanks!

I wish Github didn't say "Keno added some commits 7 days ago" for a list of commits pushed five minutes ago.

**Comment by Keno** [comment]
> Thanks!

My pleasure, as always.

> I wish Github didn't say "Keno added some commits 7 days ago" for a list of commits pushed five minutes ago.

Yeah, I believe github takes that from the commit time stamp, which doesn't update when you rebase in corrections.
